routing/payment_lifecycle.go

```diff
@@ -26,189 +28,524 @@ func (e errNoRoute) Error() string {
+	shardHandler := &shardHandler{
+		router:      p.router,
+		paymentHash: p.paymentHash,
+		quit:        make(chan struct{}),
+	}
+
+	// When the payment lifecycle loop exits, we make sure to signal any
+	// sub goroutine to exit, then wait for them to return.
+	defer func() {
+		close(shardHandler.quit)
+		shardHandler.wg.Wait()
+	}()
+
+	// Each time we send a new payment shard, we'll spin up a goroutine
+	// that will collect the result. Either a payment result will be
+	// returned, or a critical error signaling that we should immediately
+	// exit.
+	shardOutcomes := make(chan *shardOutcome)
+	criticalErrs := make(chan error)
+	var lastError error
+
+	// If we had any existing attempts outstanding, we'll start by and
+	// resume spinning up goroutines that'll collect their results and
+	// deliver them to the lifecycle loop below.
+	existingAttempts, err := p.router.cfg.Control.GetAttempts(
+		p.paymentHash,
+	)
+	if err != nil {
+		return [32]byte{}, nil, err
+	}
+
+	for _, a := range existingAttempts {
+		attempt := a
+
+		shardHandler.wg.Add(1)
+		go func() {
+			defer shardHandler.wg.Done()
+
+			outcome, err := shardHandler.collectResult(
+				&attempt.HTLCAttemptInfo,
+			)
+			if err != nil {
+				select {
+				case criticalErrs <- err:
+				case <-shardHandler.quit:
+				}
+				return
+			}
+
+			select {
+			case shardOutcomes <- outcome:
+			case <-shardHandler.quit:
+			}
+		}()
+	}
+
+	var (
+		success         = false
+		routeFailure    *paymentFailure
+		terminalFailure *paymentFailure
+	)
+
 	// We'll continue until either our payment succeeds, or we encounter a
 	// critical error during path finding.
+Loop:
 	for {
 
-		// If this payment had no existing payment attempt, we create
-		// and send one now.
-		if p.attempt == nil {
-			firstHop, htlcAdd, err := p.createNewPaymentAttempt()
-			if err != nil {
-				return [32]byte{}, nil, err
+		// Check if we got any terminal failure.
+		var failure *paymentFailure
+		switch {
+		case terminalFailure != nil:
+			failure = terminalFailure
+
+		// We only consider the route failure if we didn't have another
+		// terminal failure.
+		case routeFailure != nil:
+			failure = routeFailure
+		}
+
+		// Get all attempts made for this payment, and extract the ones
+		// still in flight.
+		attempts, err := p.router.cfg.Control.GetAttempts(
+			p.paymentHash,
+		)
+		if err != nil {
+			return [32]byte{}, nil, err
+		}
+
+		var activeShards int
+		var sentAmt lnwire.MilliSatoshi
+		for _, a := range attempts {
+			if a.Settle != nil || a.Failure != nil {
+				continue
 			}
 
-			// Now that the attempt is created and checkpointed to
-			// the DB, we send it.
-			sendErr := p.sendPaymentAttempt(firstHop, htlcAdd)
-			if sendErr != nil {
-				// TODO(joostjager): Distinguish unexpected
-				// internal errors from real send errors.
-				err = p.failAttempt(sendErr)
-				if err != nil {
-					return [32]byte{}, nil, err
-				}
+			activeShards++
+			sentAmt += a.Route.Amt()
```

Comment:
So upon resumption, it'll just attempt to send what's left? I was under the impression that some higher level logic (in Joost's follow-up PR) would handle more of this control logic...

Reply:
No splitting is done here, it just calculates what the remaining amount to send is. This value is used by the payment session to produced routes of equal or smaller value. It will probably become more clear when the follow-up PR is rebased...